STA-6133 Blog post about migrating from xstate v4 -> v5 (#390)
statelyai/docs: blog/2024-02-02-migrating-machines-to-xstate-v5/index.mdx

---
title: Migrating machines to XState v5
description: A walkthrough of migrating an existing XState v4 machine to XState v5
tags:
  [
    stately,
    studio,
    breaking changes,
    migration,
    XState v5,
    state machines,
    xstate,
  ]
authors: [kevin]
date: 2024-02-02
slug: 2024-02-02-migrating-machines-to-xstate-v5
image: /blog/2024-02-02-migrating-machines-to-xstate-v5.png
---

> We recently announced the recent release of XState v5! During its beta phase, we created a [migration guide](https://stately.ai/docs/migration), specifically to call out breaking changes and to give developers onging updates regarding api changes. This post is a walkthrough of migrating existing XState machines from v4 to v5 and is intended to be more of a step-by-step companion to the migration guide. It also focuses on migrating [XState machines that are using TypeScript](https://stately.ai/docs/typescript).

Reviewer: ```suggestion
We recently announced the recent release of XState v5! During its beta phase, we created a [migration guide](/docs/migration), specifically to call out breaking changes and to give developers onging updates regarding API changes. This post is a walkthrough of migrating existing XState machines from v4 to v5 and is intended to be more of a step-by-step companion to the migration guide. It also focuses on migrating [XState machines that are using TypeScript](/docs/typescript).
```

{/* truncate */}

:::info

Read [David’s blog post on the launch of XState v5](/blog/2023-12-01-xstate-v5).

> Prefer video? [Watch our XState v5 webinar on YouTube](https://www.youtube.com/live/TRVjeil-y74).

Reviewer: ```suggestion
Prefer video? [Watch our XState v5 webinar on YouTube](https://www.youtube.com/live/TRVjeil-y74):

<YouTube id="TRVjeil-y74" />
```

:::

We at Stately have dozens of XState machines in our own [Stately Studio](https://stately.ai/studio) code base and have been migrating those to XState v5 as well. I've learned a lot about the migration process, especially after consulting my expert teammate, Mateusz (@andarist), and I wanted to share some tips from my experience to make yours even smoother.

The following is a sequence of steps you can follow to migrate an existing XState v4 machine to XState v5. The order is merely a recommendation, not a requirement.

1. [Install XState v5 and dependencies](#1-install-xstate-v5-and-dependencies)
2. [Move types to `setup()` and remove typegen](#1-move-types-to-setup-and-remove-typegen)
3. [Convert action strings to parameterized actions](#3-convert-actions-strings-to-parameterized-actions)
4. [Include implementations or stubs to `setup()`](#4-move-implementations-or-stubs-to-setup)
5. [Provide concrete implementations](#5-provide-concrete-implementations)

> ### 1. Install XState v5 and dependencies

Reviewer: ```suggestion
## 1. Install XState v5 and dependencies
```

> The first step will be to [install XState v5](https://stately.ai/docs/xstate#installation). You can also install library-specific dependencies if you're using XState with [React](https://stately.ai/docs/xstate-react), [Vue](https://stately.ai/docs/xstate-vue), or [Svelte](https://stately.ai/docs/xstate-svelte).

Reviewer: ```suggestion
The first step will be to [install XState v5](/docs/xstate#installation). You can also install library-specific dependencies if you're using XState with [React](/docs/xstate-react), [Vue](/docs/xstate-vue), or [Svelte](/docs/xstate-svelte).
```

> If you have many state machines and would like to incrementally migrate those to v5, you can install both v4 and v5 side-by-side by [following the steps listed here](https://stately.ai/docs/migration#how-can-i-use-both-xstate-v4-and-v5). Once you've completed those steps, you'll end up with both dependencies in your `package.json`. Here's an example with libraries pegged to the latest version at the time of this writing but you can also use `npm:xstate@latest` if you prefer.

Reviewer: ```suggestion
If you have many state machines and would like to incrementally migrate those to v5, you can install both v4 and v5 side-by-side by [following the migration steps listed here](/docs/migration#how-can-i-use-both-xstate-v4-and-v5). Once you've completed those steps, you'll end up with both dependencies in your `package.json`. Here's an example with libraries pegged to the latest version at the time of this writing but you can also use `npm:xstate@latest` if you prefer.
```

Just to make the link understandable out of context.

```json5
// package.json after completing the steps in the migration guide
{
  dependencies: {
    xstate: '4.38.2',
    xstate5: 'npm:xstate@5.6.0',
    '@xstate/react': '3.2.2',
    '@xstate/react5': 'npm:@xstate/react@4.0.3',
  },
}
```

:::tip

Alternatively, you can also default `xstate` to v5 and use `xstate4` for v4.

:::

> ### 2. Move types to `setup()` and remove typegen

Reviewer: ```suggestion
## 2. Move types to `setup()` and remove typegen
```

> One of the first migration steps, based on a breaking change, is to remove any TypeScript types in `schema` from v4. These types should then be [included under a new `types` property](https://stately.ai/docs/typescript#specifying-types) in the object passed to the new `setup()` function.

Reviewer: ```suggestion
One of the first migration steps, based on a breaking change, is to remove any TypeScript types in `schema` from v4. These types should then be [included under a new `types` property](/docs/typescript#specifying-types) in the object passed to the new `setup()` function.
```

<Tabs>
<TabItem value="v5" label="XState v5">

```ts
import { setup } from 'xstate';

const machine = setup({
  types: {} as {
    context: {
      /* type context props */
    },
    events: {
      /* type events */
    },
    // Type actors here
    input:  {
      nameOfActorSource1: {
        output: {
          prop1: string;
          prop2: number;
          /* etc */
        };
      };
      nameOfActorSource2: {
        /* type actor */
      };
      nameOfActorSource3: {
        /* type actor */
      };
    },
  }
  },
  /* implementations */
  actions: {},
  guards: {},
  actors: {},
}).createMachine({
  /* machine config */
});
```

</TabItem>

<TabItem value="v4" label="XState v4">

```ts
// ❌ DEPRECATED
import { createMachine } from 'xstate';

const machine = createMachine({
  ...
  tsTypes: {} as import('./myMachine.typegen').Typegen0,
  schema: {
    context: {} as {
      /* type context props */
    },
    events: {} as {
      /* type events */
    },
    services: {} as {
      /* type services */
    },
  }
});
```

</TabItem>
</Tabs>

:::warning

> You can still pass `types` to `createMachine()` if you are using that on its own and if you prefer. However, if you chain `setup().createMachine()` then you must pass `types` to the call to `setup({ types: {} })`.

Author: ```suggestion
When chaining `setup().createMachine()` in v5, `types` can only be defined on the object passed to `setup()` and not on the config object passed to `.createMachine()`. Example: `setup({ types: {} }).createMachine({})`.
```

Author: @laurakalbag Does this sound better?

> You can still pass `types` to `createMachine()` if you are using that on its own and if you prefer. However, if you chain `setup().createMachine()` then you must pass `types` to the call to `setup({ types: {} })`.

Reviewer: ```suggestion
You can still pass `types` to `createMachine()` if you are using that on its own and if you prefer. However, if you chain `setup().createMachine()` then you must pass `types` to the call to `setup({ types: {} })`.
```

By *that*, do you mean `types` or `createMachine()`?

Author: My understanding from @Andarist is that if you use `setup()` and then use the `.createMachine` returned from that function call, `setup().createMachine()` then that `.createMachine()` does not accept `types` and you are required to pass them into the `setup()`. I'll try to clarify this in the blog text. Thanks.

:::

If you were previously using typegen in v4, you can also delete `tsTypes` as there is no more typegen in v5.

> ### 3. Convert actions strings to parameterized actions

Reviewer: ```suggestion
## 3. Convert actions strings to parameterized actions
```

> We must now [convert any action name strings to action objects](https://stately.ai/docs/migration#use-params-to-pass-params-to-actions--guards) which allows us to define an explicit `params` object (or primitive value) that will be received at runtime by our action implementations. There are two ways to do this:

Reviewer: ```suggestion
We must now [convert any action name strings to action objects](/docs/migration#use-params-to-pass-params-to-actions--guards) which allows us to define an explicit `params` object (or primitive value) that will be received at runtime by our action implementations. There are two ways to do this:
```

> We can define `params` as a [static value](https://stately.ai/docs/actions#action-objects)

Reviewer: ```suggestion
We can define `params` as a [static value](/docs/actions#action-objects)
```

```ts
actions: [
  {
    // Action type
    type: 'track',
    // Action params
    params: { response: 'good' },
  },
],
```

> We can also use [dynamic action parameters](https://stately.ai/docs/actions#dynamic-action-parameters), a function that receives `context` and/or `event`, allowing you to map those values to the `params` object.

Reviewer: ```suggestion
We can also use [dynamic action parameters](/docs/actions#dynamic-action-parameters), a function that receives `context` and/or `event`, allowing you to map those values to the `params` object.
```

        },
        {
          type: 'doThat',
          // Later, the action implementation function will be
          // passed this number value as the 2nd arg, params.
          params: ({ event }) => event.prop2,
        },
      ],
    },
  },
  states: {
    first: {
      entry: {
        type: 'whenEntering',
        // Later, the action implementation function will be
        // passed only these 2 values as the 2nd arg, params.
        params: ({ event }) => ({
          prop1: event.prop1,
          prop2: event.prop2,
        }),
      },
      exit: {
        type: 'whenExiting',
        // Later, the action implementation function will be
        // passed only these 2 values as the 2nd arg, params.
        params: ({ event }) => ({
          prop2: event.prop2,
          prop3: event.prop3,
        }),
      },
    },
  },
});
```

</TabItem>

<TabItem value="v4" label="XState v4">

```ts
// ❌ DEPRECATED
import { createMachine } from 'xstate';

const machine = createMachine({
  on: {
    next: {
      actions: ['doThis', 'doThat'],
    },
  },
  states: {
    first: {
      entry: 'whenEntering',
      exit: 'whenExiting',
    },
  },
});
```

</TabItem>
</Tabs>

> ### 4. Include implementations or stubs to `setup()`

Reviewer: ```suggestion
## 4. Include implementations or stubs to `setup()`
```

> We must [provide implementations for actions, guards, and actors](https://stately.ai/docs/machines#providing-implementations) in the object passed to `setup()`. If the machine has everything it needs to carry out these implementations on its own then these will be the actual implementations. However, if the machine needs to receive dependencies from the outside world, then these will be stubs that will be overridden later. You may end up passing a combination of concrete implementations and stubs to `setup()` which is perfectly fine.

Reviewer: ```suggestion
We must [provide implementations for actions, guards, and actors](/docs/machines#providing-implementations) in the object passed to `setup()`. If the machine has everything it needs to carry out these implementations on its own then these will be the actual implementations. However, if the machine needs to receive dependencies from the outside world, then these will be stubs that will be overridden later. You may end up passing a combination of concrete implementations and stubs to `setup()` which is perfectly fine.
```

:::info

When using TypeScript, the implementations passed to `setup()` must include types for the 2nd `params` argument.

:::

> #### Stubbing actions

Reviewer: ```suggestion
### Stubbing actions
```

```ts
setup({
  actions: {
    doThis: (_, prop1: string) => {
      // Concrete implementation here
      console.log(prop1);
    },
    // Stubbed implementation
    doThat: (_, prop2: number) => {},
    whenEntering: (_, params: { prop1: string; prop2: number }) => {
      // Concrete implementation here
      console.log(prop1, prop2);
    },
    // Stubbed implementation
    whenExiting: (_, params: { prop2: number; prop3: boolean }) => {},
  },
});
```

:::info

Stubbed actions can contain an empty function body `() => {}`

:::

> #### Stubbing guards

Reviewer: ```suggestion
### Stubbing guards
```

```ts
setup({
  guards: {
    'is this ready': (_, prop1: string) => {
      // Concrete implementation here
      return prop1.length > 0;
    },
    // Stubbed implementation
    'are we there yet': (_, prop2: number) => false,
  },
});
```

:::info

Stubbed guards must return a boolean value, either `true` or `false`.

:::

> #### Stubbing actors

Reviewer: ```suggestion
### Stubbing actors
```

Invoked actors can be stubbed with the actual actor creator helper function that will be used in the real implementation. The main purpose here is to type the `input` and `output` of the actor. There are two ways to do this:

This first way is more transferrable across other types of logic creators.

```ts
setup({
  actors: {
    doSomethingAsync: fromPromise(
      async (_: {
        input: {
          inputProp1: string;
          inputProp2: number;
        };
      }): Promise<Machine> => {
        throw new Error('Not implemented');
      },
    ),
  },
});
```

:::warning

Stubbed actors must throw an error as a way of enforcing that they will be overridden later.

:::

The second way is a bit shorter and more specific to `fromPromise`.

```ts
setup({
  actors: {
    doSomethingAsync: fromPromise<
      // Promise-wrapped output
      Machine,
      // input
      {
        inputProp1: string;
        inputProp2: number;
      }
    >(async () => {
      throw new Error('Not implemented');
    }),
  },
});
```

> #### Reducing external dependencies

Reviewer: ```suggestion
### Reducing external dependencies
```

Even dependencies that are external to the machine can be made available to the machine using one of the following methods:

> **Injecting external dependencies with `input`**

Reviewer: ```suggestion
#### Injecting external dependencies with `input`
```

If dependencies are not expected to change throughout the lifetime of the machine, then you can pass them as `input` to `setup()` and they will become available from within the machine.

```ts
setup({
  input: {
    externalDependency1: someRef,
    externalDependency2: anotherRef,
  },
});
```

> **Injecting external dependencies by sending an event**

Reviewer: ```suggestion
#### Injecting external dependencies by sending an event
```

If dependencies are expected to change over time, then you can send an event to the machine, as often as necessary (whenever dependencies change), to update the machine with the latest dependencies. For example, an event containing refs to these dependencies can be used by registered actions or actors or they can be stored in context using an `assign` action for future use.

```ts
send({
  type: 'refs.inject',
  externalDependency1,
  externalDependency2,
});
```

However, in some cases it's just not feasible or convenient to inject dependencies. There may be too many dependencies or perhaps you seek to avoid tightly coupling the machine to those dependencies. The next section describes how to provide concrete implementations to override the stubbed implementations, on a per-use basis throughout your application.

> ### 5. Provide concrete implementations

Reviewer: ```suggestion
## 5. Provide concrete implementations
```

> Stately Studio is a NextJS application so we are using the [@xstate/react](https://stately.ai/docs/xstate-react) package in our React components. We can provide concrete implementations to our stubs by using the `useActorRef()` hook. This hook allows us to pass in a machine and receive an actor ref that we can use to send events to that machine. We can provide dependencies to the machine such as our concrete implementations.

Reviewer: ```suggestion
Stately Studio is a NextJS application so we are using the [@xstate/react](/docs/xstate-react) package in our React components. We can provide concrete implementations to our stubs by using the `useActorRef()` hook. This hook allows us to pass in a machine and receive an actor ref that we can use to send events to that machine. We can provide dependencies to the machine such as our concrete implementations.
```

:::

In other components we may be using a context provider to give access to an actor ref at various levels in the component tree. A machine context provider can be created using a machine:

```ts
import { createMachine } from './machine';
import { createActorContext } from '@xstate/react';

const machine = setup({
  /* setup config */
}).createMachine({
  /* machine config */
});

export const MachineContext = createActorContext(machine);
```

It can then be imported and used in a component tree:

```tsx
import { MachineContext } from './machine';

function App() {
  return (
    <MachineContext.Provider
      logic={machine.provide({
        actions: {
          doThat: (_, prop2) => {
            // Concrete implementation here
            console.log(prop2);
          },
          whenExiting: (_, params) => {
            // Concrete implementation here
            console.log(params.prop2, params.prop3);
          },
        },
      })}
    >
      {children}
    </MachineContext.Provider>
  );
}
```

> In this case, the [provider is passed a `logic` prop](https://stately.ai/docs/xstate-react#createactorcontextlogic) whose value is the machine with the provided implementations. All the way down the component tree, we can use the [`useActorRef()` hook](https://stately.ai/docs/xstate-react#useactorrefmachine-options) to access the actor ref and send events to the machine that is using these concrete implementations.

Reviewer: ```suggestion
In this case, the [provider is passed a `logic` prop](/docs/xstate-react#createactorcontextlogic) whose value is the machine with the provided implementations. All the way down the component tree, we can use the [`useActorRef()` hook](/docs/xstate-react#useactorrefmachine-options) to access the actor ref and send events to the machine that is using these concrete implementations.
```

:::info

These implementation functions do not need to be explicitly typed again here because the `params` argument passed to them has already been typed by [implementations passed into the `setup()` function in the previous step](#4-move-implementations-or-stubs-to-setup).

:::

> #### Providing actor implementations

Reviewer: ```suggestion
### Providing actor implementations
```

This is a bit more involved as we need to define 3 things:

1. Provide either a [concrete implementation or a stub for the actor creator function in `setup()`](#stubbing-actors)
2. Register the `invoke` in a state in the main machine config.
3. Provide a concrete implementation for the actor if not already passing it to `setup()`.

> **Registering the `invoke` in a state in the main machine config**

Reviewer: ```suggestion
#### Registering the `invoke` in a state in the main machine config
```

This is similar to how invoked actors were registered in v4. The main difference is we also define the `input` here where we map event values to input values. The actions registered with `onDone` and `onError` are also defined as objects with `params` just like [we saw before with transition actions and entry/exit actions](#3-convert-actions-strings-to-parameterized-actions).

```ts
import { createMachine } from 'xstate';

createMachine({
  /* machine config */
  states: {
    /* other states */
    someState: {
      invoke: {
        src: 'doSomethingAsync', // required
        id: 'doSomethingAsync', // optional
        input: ({ event }) => ({
          inputProp1: event.prop1,
          inputProp2: event.prop2,
        }),
        onDone: {
          target: 'Idle',
          actions: [
            {
              type: 'showSuccessToast',
            },
            {
              type: 'handleOutputOnSuccess',
              params: ({ event }) => event.output,
            },
          ],
        },
        onError: {
          target: 'Idle',
          actions: [{ type: 'showErrorToast' }],
        },
      },
    },
  },
});
```

> **Providing a concrete implementation for the actor if not already passing it to `setup()`**

Reviewer: ```suggestion
#### Providing a concrete implementation for the actor if not already passing it to `setup()`
```

If we only defined a stub for the actor creator function in `setup()` then we must provide a concrete implementation for the actor with our other concrete implementations.

```ts
import { useActorRef } from '@xstate/react';

const actorRef = useActorRef(
  machine.provide({
    actors: {
      doSomethingAsync: fromPromise(
        ({
          input,
        }: {
          input: {
            inputProp1: string;
            inputProp2: number;
          };
        }) => {
          return trpcProxyClient.stuff.asyncstuff.mutate(input);
        },
      ),
    },
  }),
);
```

> ### Summary

Reviewer: ```suggestion
## Summary
```

By following the sequence of steps above, you should be able to migrate your existing XState v4 machines to XState v5 and get all of the types working with little to no need for type narrowing via assertions or type guards.

> I hope this walkthrough has been helpful in migrating your existing XState v4 machines to XState v5. If you have any questions or feedback, please [reach out to us on [our Discord](https://discord.gg/xstate).

Reviewer: ```suggestion
I hope this walkthrough has been helpful in migrating your existing XState v4 machines to XState v5. If you have any questions or feedback, please [reach out to us on our Discord](https://discord.gg/xstate).
```

---
title: Migrating machines to XState v5
description: A walkthrough of migrating an existing XState v4 machine to XState v5
tags:
  [
    stately,
    studio,
    breaking changes,
    migration,
    XState v5,
    state machines,
    xstate,
  ]
authors: [kevin]
date: 2024-02-02
slug: 2024-02-02-migrating-machines-to-xstate-v5
image: /blog/2024-02-02-migrating-machines-to-xstate-v5.png
---

We recently announced the recent release of XState v5! During its beta phase, we created a [migration guide](/docs/migration), specifically to call out breaking changes and to give developers onging updates regarding API changes. This post is a walkthrough of migrating existing XState machines from v4 to v5 and is intended to be more of a step-by-step companion to the migration guide. It also focuses on migrating [XState machines that are using TypeScript](/docs/typescript).

{/* truncate */}

:::info

Read [David’s blog post on the launch of XState v5](/blog/2023-12-01-xstate-v5).

Prefer video? [Watch our XState v5 webinar on YouTube](https://www.youtube.com/live/TRVjeil-y74):

<YouTube id="TRVjeil-y74" />

:::

> We at Stately have dozens of XState machines in our own [Stately Studio](https://stately.ai/studio) code base and have been migrating those to XState v5 as well. I've learned a lot about the migration process, especially after consulting my expert teammate, Mateusz (@andarist), and I wanted to share some tips from my experience to make yours even smoother.

Author: @Andarist Should I link to your GitHub or just mention you by real name only? What do you like here?

```json5
// package.json after completing the steps in the migration guide
{
  dependencies: {
    xstate: '4.38.2',
    xstate5: 'npm:xstate@5.6.0',
    '@xstate/react': '3.2.2',
    '@xstate/react5': 'npm:@xstate/react@4.0.3',
  },
}
```

:::tip

Alternatively, you can also default `xstate` to v5 and use `xstate4` for v4.

:::

## 2. Move types to `setup()` and remove typegen

One of the first migration steps, based on a breaking change, is to remove any TypeScript types in `schema` from v4. These types should then be [included under a new `types` property](/docs/typescript#specifying-types) in the object passed to the new `setup()` function.

<Tabs>
<TabItem value="v5" label="XState v5">

```ts
import { setup } from 'xstate';

const machine = setup({
  types: {} as {
    context: {
      /* type context props */
    },
    events: {
      /* type events */
    },
    // Type actors here
    input:  {
      nameOfActorSource1: {
        output: {
          prop1: string;
          prop2: number;
          /* etc */
        };
      };
      nameOfActorSource2: {
        /* type actor */
      };
      nameOfActorSource3: {
        /* type actor */
      };

>     },

Reviewer 2: This is a bit confusing; the actors aren't meant to be typed here

          prop2: event.prop2,
          prop3: event.prop3,
        }),
      },
    },
  },
});
```

</TabItem>

<TabItem value="v4" label="XState v4">

```ts
// ❌ DEPRECATED
import { createMachine } from 'xstate';

const machine = createMachine({
  on: {
    next: {
      actions: ['doThis', 'doThat'],
    },
  },
  states: {
    first: {
      entry: 'whenEntering',
      exit: 'whenExiting',
    },
  },
});
```

</TabItem>
</Tabs>

## 4. Include implementations or stubs in `setup()`

We must [provide implementations for actions, guards, and actors](/docs/machines#providing-implementations) in the object passed to `setup()`. If the machine has everything it needs to carry out these implementations on its own then these will be the actual implementations. However, if the machine needs to receive dependencies from the outside world, then these will be stubs that will be overridden later. You may end up passing a combination of concrete implementations and stubs to `setup()` which is perfectly fine.

:::info

When using TypeScript, the implementations passed to `setup()` must include types for the 2nd `params` argument.

:::

### Stubbing actions

```ts
setup({
  actions: {

>     doThis: (_, prop1: string) => {
>       // Concrete implementation here
>       console.log(prop1);
>     },
>     // Stubbed implementation
>     doThat: (_, prop2: number) => {},

Reviewer 2: ```suggestion
    doThis: (_, params: string) => {
      // Concrete implementation here
      console.log(params);
    },
    // Stubbed implementation
    doThat: (_, params: number) => {},
```

    whenEntering: (_, params: { prop1: string; prop2: number }) => {
      // Concrete implementation here
      console.log(prop1, prop2);
    },
    // Stubbed implementation
    whenExiting: (_, params: { prop2: number; prop3: boolean }) => {},
  },
});
```

:::info

Stubbed actions should use an empty function body `() => {}`

:::

### Stubbing guards

```ts
setup({
  guards: {

>     'is this ready': (_, prop1: string) => {
>       // Concrete implementation here
>       return prop1.length > 0;
>     },
>     // Stubbed implementation
>     'are we there yet': (_, prop2: number) => false,

Reviewer 2: ```suggestion
    'is this ready': (_, params: string) => {
      // Concrete implementation here
      return params.length > 0;
    },
    // Stubbed implementation
    'are we there yet': (_, params: number) => false,
```

I also recommend always using objects for params; more explicit

Author: I went back and forth with that. Should we recommend that users always use an object named `params` here for consistency (unless they destructure in which case there's no name)?

  },
});
```

:::info

Stubbed guards must return a boolean value, either `true` or `false`.

:::

### Stubbing actors

Invoked actors can be stubbed with the actual actor creator helper function that will be used in the real implementation. The main purpose here is to type the `input` and `output` of the actor. There are two ways to do this:

This first way is more transferrable across other types of logic creators.

```ts
setup({
  actors: {
    doSomethingAsync: fromPromise(
      async (_: {
        input: {
          inputProp1: string;
          inputProp2: number;
        };
      }): Promise<Machine> => {
        throw new Error('Not implemented');
      },
    ),
  },
});
```

:::warning

Stubbed actors must throw an error as a way of enforcing that they will be overridden later.

:::

The second way is a bit shorter and more specific to `fromPromise`.

```ts
setup({
  actors: {
    doSomethingAsync: fromPromise<
      // Promise-wrapped output

>       Machine,

Author: Good catch. That's taken from a real Studio machine so I'll change the name to be less confusing in this example.

5. [Provide concrete implementations](#5-provide-concrete-implementations)

## 1. Install XState v5 and dependencies

The first step will be to [install XState v5](/docs/xstate#installation). You can also install the library-specific dependencies that are compatible with XState v5 if you’re using [React](/docs/xstate-react), [Vue](/docs/xstate-vue), or [Svelte](/docs/xstate-svelte).

If you have many state machines and would like to incrementally migrate those to v5, you can install both v4 and v5 side-by-side by [following the migration steps listed here](/docs/migration#how-can-i-use-both-xstate-v4-and-v5). Once you’ve completed those steps, you’ll end up with both dependencies in your `package.json`. Here’s an example with libraries pegged to the latest version at the time of this writing but you can also use `npm:xstate@latest` if you prefer.

:::tip

Alternatively, you can also default `xstate` to v5 and use `xstate4` for v4.

:::

```json5
// package.json after completing the steps in the migration guide
{
  dependencies: {
    xstate: '4.38.2',
    xstate5: 'npm:xstate@5.6.0',
    '@xstate/react': '3.2.2',

>     '@xstate/react5': 'npm:@xstate/react@4.0.3',

Author: I actually linked to the section in the Migration Doc that shows these details but I should mention that there's a script so users know why they should consult the full version. Or maybe I'll remove the example from this blog post and only link to the full version in the docs so it's less "incomplete" and simply a referral with commentary.

:::warning

Calling `setup().createMachine()` from XState v5 but then calling `assign` from v4 within a v5 machine could result in type errors.

:::

## 2. Move types to `setup()` and remove typegen

One of the first migration steps, based on a breaking change, is to remove any TypeScript types from the `schema` property of the machine config from v4. These types should now be [included under a new `types` property](/docs/typescript#specifying-types) in the object passed to the new `setup({})` function.

<Tabs>
<TabItem value="v5" label="XState v5">

```ts
import { setup } from 'xstate';

const machine = setup({
  types: {} as {
    context: {
      /* type context props */
    };
    events: {
      /* type events */
    };
  },
  /* implementations */
  actions: {},
  guards: {},
  actors: {},
}).createMachine({
  /* machine config */
});
```

</TabItem>

<TabItem value="v4" label="XState v4">

```ts
// ❌ DEPRECATED

> import { createMachine } from 'xstate';

Reviewer 3: It's not exactly deprecated - we still have this API and we didn't deprecate it. `setup` is preferred though since it automates converting implementations for actions (and more) to types.

Author: Great. Thanks for this clarification. I'll remove "deprecated" and add mention of the reasoning as well.

const machine = createMachine({
  ...
  tsTypes: {} as import('./myMachine.typegen').Typegen0,
  schema: {
    context: {} as {
      /* type context props */
    },
    events: {} as {
      /* type events */
    },
    services: {} as {
      /* type services */
    },
  }
});
```

</TabItem>
</Tabs>

When defining types, you may have noticed in past examples that an empty object is often cast as the types you want for each property like `events` and `actions`.

```ts
const machine = setup({
  types: {
    context: {} as {
    };
    events: {} as {
    };
  },
})
```

> That still works but but it’s even easier to just cast the whole types object all at once.

Reviewer 3: Yeah, this isn't actually anything new - it worked the same way in v4. But I endorse an extra mention of this here since I think people often miss this

Author: Makes sense.

```ts
const machine = setup({
  types: {} as {
    context: {};

>     events: {};

Reviewer 3: this is likely a type error because it doesn't adhere to the constraint
```suggestion
    events: { type: 'EVENT' };
```

Author: Thanks for catching that.

  },
});
```

:::warning

When chaining `setup().createMachine()` in v5, `types` can only be defined on the object passed to `setup()` and not on the config object passed to `.createMachine()`. Example: `setup({ types: {} }).createMachine({})`. A chained call to `createMachine()` only accepts the one argument, the machine config.

:::

### Remove typegen

If you were previously using typegen in v4, you can also delete `tsTypes` as there is no more typegen in v5. Typing events in actions and guards can be done using type narrowing in those implementations. However, the following sections will show you how you can type a new `params` argument passed to those implementations.

## 3. Convert action and guard strings to parameterized objects

> We must now [convert any named action or guard strings to action objects](/docs/migration#use-params-to-pass-params-to-actions--guards) which allows us to define an explicit `params` object (or primitive value) that will be received at runtime by our action implementations. There are two ways to do this:

Reviewer 3: That is only a "requirement" when you depend on `event` because it helps you to adapt `event` into `params` within transition actions. If you have simple actions - or ones that only depend on `context` then you can continue using `entry: 'doStuff'` etc. But with `setup` you still have to declare `doStuff` in it.

Author: I'll definitely add some of the more nuanced considerations here and will make this a non "requirement".

We can define `params` as a [static value](/docs/actions#action-objects)

```ts
actions: [
  {
    // Action type
    type: 'track',
    // Action params
    params: { response: 'good' },
  },
],
```

We can also use [dynamic action parameters](/docs/actions#dynamic-action-parameters), a function that receives `context` and/or `event`, allowing you to map those values to the `params` object.

```ts
entry: [
  {
    type: 'logInitialRating',
    params: ({ context }) => ({
      initialRating: context.initialRating,
    }),
  },
];
```

Here is a more complete example of converting action strings to parameterized actions:

<Tabs>
<TabItem value="v5" label="XState v5">

```ts
import { setup } from 'xstate';

const machine = setup({

>   types: {
>     events: {} as {
>       type: 'next';
>       prop1: string;
>       prop2: number;
>       prop3: boolean;
>     },
>   },

Reviewer 3: ```suggestion
  types: {} as {
    events: {
      type: 'next';
      prop1: string;
      prop2: number;
      prop3: boolean;
    },
  },
```

  /* more setup */
}).createMachine({
  on: {
    next: {
      target: 'first',
      actions: [
        {
          guard: {
            type: 'is this ready',
            params: ({ event }) => ({ ready: event.ready }),
          },
          type: 'doThis',
          // Later, the action implementation function will be
          // passed this string value in the 2nd arg, params.
          params: ({ event }) => ({ prop1: event.prop1 }),
        },
        {
          type: 'doThat',
          // Later, the action implementation function will be
          // passed this number value in the 2nd arg, params.
          params: ({ event }) => ({ prop2: event.prop2 }),
        },
      ],
    },
  },
  states: {
    first: {

>       entry: {

Reviewer 3: this totally works but the downside is that within `entry` and `exit` we can't  narrow `event` automatically, so parameterized actions don't give us a lot here if we are after converting `event` to `params` for the improved type experience. It might still be a good pattern to have everything to use the same pattern etc but the biggest benefits of `params` are seen when used with transition actions

Author: Thanks for the clarification. I'll definitely mention the difference between `entry` and `exit` actions and transition actions here.

We must [provide implementations for actions, guards, and actors](/docs/machines#providing-implementations) in the object passed to `setup()`. If the machine has everything it needs to carry out these implementations on its own then these will be the actual implementations. However, if the machine needs to receive dependencies from the outside world, then these will be stubs that will be overridden later. You may end up passing a combination of concrete implementations and stubs to `setup()` which is perfectly fine.

:::warning

When using TypeScript, it is necessary to type the 2nd `params` argument passed to each implementation function.

:::

### Stubbing actions

```ts
setup({
  actions: {
    doThis: (_, params: { prop: string }) => {
      // Concrete implementation here
      console.log(params.prop);
    },
    // Stubbed implementation
    doThat: (_, params: { prop: number }) => {},
    whenEntering: (_, params: { prop1: string; prop2: number }) => {
      // Concrete implementation here
      console.log(prop1, prop2);
    },
    // Stubbed implementation
    whenExiting: (_, params: { prop2: number; prop3: boolean }) => {},
  },
});
```

:::info

> Stubbed actions should use an empty function body `() => {}`

Author: I like the idea of throwing in a stubbed action by default or as a more explicit alternative to an empty body`{}`.

@Andarist Can we also throw in a stubbed guard or is it better to just return a boolean and override it in a concrete implementation?
